wordpress.org/public_html/wp-content/plugins/wporg-gp-translation-suggestions/inc/routes/class-translation-memory.php

```diff
@@ -31,13 +32,196 @@ public function get_suggestions( $project_path, $locale_slug, $set_slug ) {
+				// Use the translation as key, because we could have multiple translations with the same term.
+				$glossary_entries[ $gp_glossary_entry->translation ] = $gp_glossary_entry->term;
+			}
+		}
+		if ( ! empty( $glossary_entries ) ) {
+			$glossary_query = ' The following terms are translated as follows: ';
+			foreach ( $glossary_entries as $translation => $term ) {
+				$glossary_query .= '"' . $term . '" is translated as "' . $translation . '"';
+				if ( array_key_last( $glossary_entries ) != $translation ) {
+					$glossary_query .= ', ';
+				}
+			}
+			$glossary_query .= '.';
+		}
+
+		$gp_locale     = GP_Locales::by_field( 'slug', $locale );
+		$openai_query .= ' Translate the following text to ' . $gp_locale->english_name . ": \n";
+		$openai_query .= '"' . $original_singular . '"';
+
+		$messages = array(
+			array(
+				'role'    => 'system',
+				'content' => $openai_prompt . $glossary_query,
+			),
+			array(
+				'role'    => 'user',
+				'content' => $openai_query,
+			),
+		);
+
+		$openai_response = wp_remote_post(
+			'https://api.openai.com/v1/chat/completions',
+			array(
+				'timeout' => 20,
+				'headers' => array(
+					'Content-Type'  => 'application/json',
+					'Authorization' => 'Bearer ' . $openai_key,
+				),
+				'body'    => wp_json_encode(
+					array(
+						'model'       => 'gpt-3.5-turbo',
+						'max_tokens'  => 1000,
+						'n'           => 1,
+						'messages'    => $messages,
+						'temperature' => $openai_temperature,
+					)
+				),
+			)
+		);
+		if ( is_wp_error( $openai_response ) ) {
+			return array();
+		}
+		$response_status = wp_remote_retrieve_response_code( $openai_response );
+		if ( 200 !== $response_status ) {
+			return array();
+		}
+		$output  = json_decode( wp_remote_retrieve_body( $openai_response ), true );
+		$message = $output['choices'][0]['message'];
+		$response['openai']['translation'] = trim( trim( $message['content'] ), '"' );
+		$response['openai']['diff'] = '';
+
+		return $response;
+	}
+
+	/**
+	 * Gets a translation suggestion from DeepL.
+	 *
+	 * @param string $original_singular The singular from the original string.
+	 * @param string $locale            The locale.
+	 *
+	 * @return array
+	 */
+	private function get_deepl_suggestion( string $original_singular, string $locale ): array {
+		$free_url        = 'https://api-free.deepl.com/v2/translate';
+		$gp_default_sort = get_user_option( 'gp_default_sort' );
+		$deepl_api_key   = gp_array_get( $gp_default_sort, 'deepl_api_key' );
+		if ( empty( trim( $deepl_api_key ) ) ) {
+			return array();
+		}
+		$source_lang = 'EN';
+		$target_lang = $this->get_deepl_locale( $locale );
+		if ( empty( $target_lang ) ) {
+			return array();
+		}
+		$deepl_response = wp_remote_post(
+			$free_url,
+			array(
+				'timeout' => 20,
+				'body'    => array(
+					'auth_key'    => $deepl_api_key,
+					'text'        => $original_singular,
+					'source_lang' => $source_lang,
+					'target_lang' => $target_lang,
```

Comment:
Could we extend the mapping so that it can also include support [for the `formality` parameter](https://www.deepl.com/docs-api/translate-text/translate-text/)? Without it, it is unusable in German since we use informal German in WordPress but DeepL creates formal German.

Reply:
Done! I have tested with Spanish (Spain) and works fine! Thank you for the suggestion. 

![image](https://user-images.githubusercontent.com/1667814/226709532-eca44f12-2d99-4df4-9c25-38b44577a871.png)

Reply:
I have moved the select item for the language formality from the user preferences to a method with the values hard-coded for each language. I have [pinged](https://wordpress.slack.com/archives/C02RP50LK/p1679476128686079) to the GTE for each target language supported in Deepl, so this array should be updated as we receive information from the GTEs.

![image](https://user-images.githubusercontent.com/1667814/226873647-fcca60c6-f716-49ad-9ba5-bc6ee70ddcac.png)